### PR TITLE
Auto Case Update Refactor

### PR DESCRIPTION
**Pull request by gcapalbo**
Details here:
https://docs.google.com/document/d/1M3OWqYyTQKBGvB_xEmkB9bsTH1ZvCGNXqIs3nvRQxR0/edit#

Demo here:
https://staging.commcarehq.org/a/gc-staging/data/edit/automatic_updates/

buddy @mkangia 
cc @calellowitz for updates to rule running processes (this extends the work you had done)

### CODE REVIEW

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
+/* globals ko, $ */
+
+var CaseRuleActions = function(initial) {
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js]: 'CaseRuleActions' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            var obj = new CustomActionDefinition();
+            obj.name(value.name);
+            self.actions.push(obj);
+        });
+    };
+};
+
+var CloseCaseDefinition = function() {
+    'use strict';
+    var self = this;
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js]: 'self' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
+/* globals ko, $ */
+
+var CaseRuleCriteria = function(initial, constants) {
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js]: 'CaseRuleCriteria' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                obj.property_name(value.property_name);
+                obj.property_value(value.property_value);
+                obj.match_type(value.match_type);
+                self.criteria.push(obj);
+            } else if(
+                value.match_type === constants.MATCH_DAYS_BEFORE ||
+                value.match_type === constants.MATCH_DAYS_AFTER
+            ) {
+                var days = Number.parseInt(value.property_value);
+                var obj;
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js]: 'obj' is already defined (no-redeclare)

**Comment by gcapalbo** [comment]
fine to review, but i want to add tests before merging

**Review by mkangia — APPROVED** [review]
I don't have any context on this and would take quite a lot of digging in. Will defer to @calellowitz  for review or if you want to add some more reviewers @gcapalbo

File: corehq/apps/data_interfaces/templates/data_interfaces/edit_case_rule.html [hunk 5]
```diff
@@ -19,6 +20,14 @@
     }, {{ criteria_form.constants|JSON }});
     $('#rule-criteria').koApplyBindings(criteria_model);
     criteria_model.load_initial();
+
+    var actions_model = new CaseRuleActions({
+        close_case: '{{ actions_form.close_case.value|default:"false"|escapejs }}',
+        properties_to_update: {{ actions_form.properties_to_update.value|default:"[]"|safe }},
+        custom_action_definitions: {{ actions_form.custom_action_definitions.value|default:"[]"|safe }},
```
Comment (calellowitz) [inline corehq/apps/data_interfaces/templates/data_interfaces/edit_case_rule.html]: same comment as above

**Review by calellowitz — CHANGES_REQUESTED** [review]
left one real change request, hopefully it came through and isnt lost to an old diff

**Comment by calellowitz** [comment]
Looks good besides the one comment, though admittedly it was large enough of a PR that I was not able to read every part super closely. Let me know if github hides the comment since you ended up modifying the section i made it on.

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/data_interfaces/models.py, line 956 - F821 undefined name 'domain'
* corehq/apps/data_interfaces/models.py, line 964 - F821 undefined name 'since'
* corehq/apps/data_interfaces/models.py, line 982 - F821 undefined name 'FormAccessors'
* corehq/apps/data_interfaces/models.py, line 985 - F821 undefined name 'instance'
* corehq/apps/data_interfaces/tests/test_case_rule_migration.py, line 1 - F401 'json' imported but unused

File: corehq/apps/data_interfaces/tests/test_case_rule_migration.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
+import json
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_case_rule_migration.py]: F401 'json' imported but unused

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+hqDefine("data_interfaces/js/case_rule.js", function() {
+
+    $(function() {
+        if(hqImport("hqwebapp/js/initial_page_data.js").get('read_only_mode')) {
+            $('#rule-definition :input').prop('disabled', true);
+        }
+
+        $("#rule-definition-form").submit(function(event) {
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule.js]: 'event' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                var obj = new CustomActionDefinition();
+                obj.name(value.name);
+                self.actions.push(obj);
+            });
+        };
+    };
+
+    var CloseCaseDefinition = function() {
+        'use strict';
+        var self = this;
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js]: 'self' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+    $(function() {
+        actions_model = new CaseRuleActions(hqImport("hqwebapp/js/initial_page_data.js").get('actions_initial'));
+        $('#rule-actions').koApplyBindings(actions_model);
+        actions_model.load_initial();
+    });
+
+    return {
+        get_actions_model: function() {return actions_model;},
+    }
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js]: Missing semicolon. (semi)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+            hqImport("hqwebapp/js/initial_page_data.js").get('criteria_initial'),
+            hqImport("hqwebapp/js/initial_page_data.js").get('criteria_constants')
+        );
+        $('#rule-criteria').koApplyBindings(criteria_model);
+        criteria_model.load_initial();
+    });
+
+    return {
+        get_criteria_model: function() {return criteria_model;},
+    }
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js]: Missing semicolon. (semi)

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py [hunk 11]
```diff
@@ -546,3 +569,338 @@ def set_parent_case(domain, child_case, parent_case):
...
+
+            hqcase.utils.update_case(self.domain, case.case_id, case_properties={'result': 'x'})
+            case = CaseAccessors(self.domain).get_case(case.case_id)
+            self.assertFalse(rule.criteria_match(case, datetime.utcnow()))
+
+            hqcase.utils.update_case(self.domain, case.case_id, case_properties={'result': 'negative'})
+            case = CaseAccessors(self.domain).get_case(case.case_id)
+            self.assertTrue(rule.criteria_match(case, datetime.utcnow()))
+
+    @run_with_all_backends
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: F811 redefinition of unused 'test_case_property_equal' from line 619

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py [hunk 12]
```diff
@@ -546,3 +573,819 @@ def set_parent_case(domain, child_case, parent_case):
...
+                task_patch.assert_not_called()
+
+
+class CaseRuleEndToEndTests(BaseCaseRuleTest):
+
+    def test_get_rules_from_domain(self):
+        rule1 = _create_empty_rule(self.domain, case_type='person-1')
+        rule2 = _create_empty_rule(self.domain, case_type='person-1')
+        rule3 = _create_empty_rule(self.domain, case_type='person-2')
+        rule4 = _create_empty_rule(self.domain, case_type='person-2', active=False)
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: F841 local variable 'rule4' is assigned to but never used

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py [hunk 13]
```diff
@@ -546,3 +573,819 @@ def set_parent_case(domain, child_case, parent_case):
...
+        rule5 = _create_empty_rule(self.domain, case_type='person-3', deleted=True)
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: F841 local variable 'rule5' is assigned to but never used

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py [hunk 14]
```diff
@@ -546,3 +573,819 @@ def set_parent_case(domain, child_case, parent_case):
...
+        rule2.filter_on_server_modified = True
+        rule2.server_modified_boundary = 20
+        rule2.save()
+
+        rule3 = _create_empty_rule(self.domain, case_type='person-2')
+        rule3.filter_on_server_modified = True
+        rule3.server_modified_boundary = 30
+        rule3.save()
+
+        rule4 = _create_empty_rule(self.domain, case_type='person-2')
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: F841 local variable 'rule4' is assigned to but never used

**Comment by gcapalbo** [comment]
@calellowitz I externalized all the js and added some more tests - this should be good to go now.

(Leaving the don't merge label just so it doesn't get accidentally deployed overnight or something)

**Review by calellowitz — APPROVED** [review]

**Pull request merged**